pkg/engine/task/task_apply.go

```diff
@@ -99,33 +137,87 @@ func apply(rr []runtime.Object, c client.Client, di discovery.DiscoveryInterface
+	if err != nil {
+		return fmt.Errorf("failed to get modified config %v", err)
+	}
+
+	// Create the actual patch
+	var patchData []byte
+	var patchType types.PatchType
+	if useJSONMerge(updatedObj) {
+		patchType = types.MergePatchType
+		patchData, err = jsonThreeWayMergePatch(original, modified, current)
+	} else {
+		patchType = types.StrategicMergePatchType
+		patchData, err = strategicThreeWayMergePatch(updatedObj, original, modified, current)
+	}
+
+	if err != nil {
+		// TODO: We could try to delete/create here, but that would be different behavior from before
+		return fmt.Errorf("failed to create patch: %v", err)
+	}
+
+	// Execute the patch
+	err = c.Patch(context.TODO(), updatedObj, client.ConstantPatch(patchType, patchData))
+	if err != nil {
+		return fmt.Errorf("failed to execute patch: %v", err)
+	}
+	return nil
+}
+
+func useJSONMerge(newObj runtime.Object) bool {
 	_, isUnstructured := newObj.(runtime.Unstructured)
 	_, isCRD := newObj.(*apiextv1beta1.CustomResourceDefinition)
+	return isUnstructured || isCRD || isKudoType(newObj)
+}
 
-	if isUnstructured || isCRD || isKudoType(newObj) {
-		// strategic merge patch is not supported for these types, falling back to merge patch
-		err := c.Patch(context.TODO(), newObj, client.ConstantPatch(types.MergePatchType, newObjJSON))
-		if err != nil {
-			return fmt.Errorf("failed to apply merge patch to object %s/%s: %w", key.Namespace, key.Name, err)
-		}
-	} else {
-		err := c.Patch(context.TODO(), newObj, client.ConstantPatch(types.StrategicMergePatchType, newObjJSON))
-		if err != nil {
-			return fmt.Errorf("failed to apply StrategicMergePatch to object %s/%s: %w", key.Namespace, key.Name, err)
+func jsonThreeWayMergePatch(original, modified, current []byte) ([]byte, error) {
+	preconditions := []mergepatch.PreconditionFunc{
+		mergepatch.RequireKeyUnchanged("apiVersion"),
+		mergepatch.RequireKeyUnchanged("kind"),
+		mergepatch.RequireMetadataKeyUnchanged("name"),
+	}
+
+	patchData, err := jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+	if err != nil {
+		if mergepatch.IsPreconditionFailed(err) {
+			return nil, fmt.Errorf("%s", "At least one of apiVersion, kind and name was changed")
 		}
+		return nil, fmt.Errorf(" failed to create json merge patch: %v", err)
 	}
 
-	return nil
+	return patchData, nil
+}
+
+func strategicThreeWayMergePatch(r runtime.Object, original, modified, current []byte) ([]byte, error) {
+	// Create the patch
+	patchMeta, err := strategicpatch.NewPatchMetaFromStruct(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create patch meta %v", err)
+	}
+
+	patchData, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, patchMeta, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create patch data %v", err)
+	}
+
+	return patchData, nil
 }
 
 func isKudoType(object runtime.Object) bool {
```

Comment:
double-nit: could you please move this method up closer to `func shouldMergeThreeWays` since that's the only place it is currently used